### PR TITLE
change ints to enums where appropriate

### PR DESCRIPTION
**Pull request by gridley**
$TITLE, removes a few TODOs.

### CODE REVIEW

**Comment by gridley** [comment]
@paulromano Do you all still need `extern "C"` on the tentative enum which would contain BC_TRANSMIT, BC_VACUUM, etc? (or need it to remain an int constexpr, for Nek5000 Fortran I assume?) CUDA gets mad with `extern "C" enum` for some reason, so I would prefer either to leave as ints or not have extern "C".

Really strangely, an xtensor error ends up getting thrown only when compiled by mpic++ right now. Aside from that, this PR seems to be working. I'll be looking into this...

**Comment by paulromano** [comment]
@gridley No, I don't see any need for `extern "C"` on that particular enum. Thanks for checking!

File: include/openmc/cell.h [hunk 1]
```diff
@@ -23,17 +23,21 @@ namespace openmc {
...
-constexpr int FILL_UNIVERSE {2};
-constexpr int FILL_LATTICE {3};
-
-// TODO: Convert to enum
-constexpr int32_t OP_LEFT_PAREN   {std::numeric_limits<int32_t>::max()};
-constexpr int32_t OP_RIGHT_PAREN  {std::numeric_limits<int32_t>::max() - 1};
-constexpr int32_t OP_COMPLEMENT   {std::numeric_limits<int32_t>::max() - 2};
-constexpr int32_t OP_INTERSECTION {std::numeric_limits<int32_t>::max() - 3};
-constexpr int32_t OP_UNION        {std::numeric_limits<int32_t>::max() - 4};
+enum CellFillType
```
Comment (paulromano) [inline include/openmc/cell.h]: [Prefer `enum class` to plain `enum`](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-class)

File: include/openmc/cell.h [hunk 2]
```diff
@@ -23,17 +23,21 @@ namespace openmc {
...
+{
+  FILL_MATERIAL = 1,
+  FILL_UNIVERSE = 2,
+  FILL_LATTICE = 3
```
Comment (paulromano) [inline include/openmc/cell.h]: Would removing the integer values here and relying on the defaults do any harm?

File: include/openmc/cell.h (hunk 1 above)
Comment (paulromano) [reply include/openmc/cell.h]: Also, the prevailing style when defining a struct/class is to group the opening brace with the name:
```C++
enum class CellFillType {
  ...
```

File: include/openmc/constants.h [hunk 3]
```diff
@@ -114,190 +114,206 @@ constexpr int NUCLIDE_NONE  {-1};
...
-constexpr int ANGLE_TABULAR {3};
-constexpr int ANGLE_LEGENDRE {4};
-constexpr int ANGLE_HISTOGRAM {5};
+enum AngleDistributionType
+{
+  ANGLE_ISOTROPIC = 1,
+  ANGLE_32_EQUI = 2,
+  ANGLE_TABULAR = 3,
+  ANGLE_LEGENDRE = 4,
+  ANGLE_HISTOGRAM = 5,
```
Comment (paulromano) [inline include/openmc/constants.h]: Same comment regarding default values

File: include/openmc/constants.h [hunk 4]
```diff
@@ -114,190 +114,206 @@ constexpr int NUCLIDE_NONE  {-1};
...
+  LIBRARY_MULTIGROUP = 4
+};
+
+enum ProbabilityTableParameter
+{
+  URR_CUM_PROB,
+  URR_TOTAL   ,
+  URR_ELASTIC ,
+  URR_FISSION ,
+  URR_N_GAMMA ,
```
Comment (paulromano) [inline include/openmc/constants.h]: Remove spaces before commas

File: src/tallies/tally.cpp [hunk 5]
```diff
@@ -906,7 +906,7 @@ void reduce_tally_results()
     auto& tally {model::tallies[i_tally]};
 
     // Get view of accumulated tally values
-    auto values_view = xt::view(tally->results_, xt::all(), xt::all(), RESULT_VALUE);
+    auto values_view = xt::view(tally->results_, xt::all(), xt::all(), (int)RESULT_VALUE);
```
Comment (paulromano) [inline src/tallies/tally.cpp]: [Use a named cast](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-casts-named)

Comment (gridley) [reply src/tallies/tally.cpp]: Still weaning off my old C ways, thanks for this tip!

File: include/openmc/constants.h (hunk 4 above)
Comment (gridley) [reply include/openmc/constants.h]: Mannnnn, but I like it when it all lines up! :wink: (Not a problem to make that change)

File: include/openmc/cell.h (hunk 2 above)
Comment (gridley) [reply include/openmc/cell.h]: Probably not, but I didn't want to peruse the whole code base to make sure. I should look into this for the individual enums though. One example where the numeric values of the enum are actually used is in the second commit of this PR, where an explicit cast to int is required in MPI mode. So... I'll be double-checking each enum to remove this.

**Comment by gridley** [comment]
OK, so, I think leaving the MT numbers as a non enum class makes sense from now. Using an enum class requires a huge amount of extra typing, and removes implicit conversion to an integer (which makes a lot of sense when you're talking about MT numbers). It takes a ton of search and replace if you still want enum class here though @paulromano.

Although, yes, for most other stuff, enum classes make sense.

**Comment by paulromano** [comment]
@gridley That's ok. Do what you think is sensible and then we'll take it from there. Give me another ping when this is ready for review once more. :+1:

File: include/openmc/cell.h (hunk 1 above)
Comment (gridley) [reply include/openmc/cell.h]: OK, both of these things are now included in the PR.

File: include/openmc/constants.h (hunk 4 above)
Comment (gridley) [reply include/openmc/constants.h]: this change is now incorporated!

**Comment by gridley** [comment]
@paulromano So, the change to enum classes was fairly large, but I think worthwhile. Please take a look now when you have a chance. In addition, I moved a few enums defined in constants.h which are only used in one class to the class that uses them.

**Comment by pshriwise** [comment]
@gridley it looks like the enum values for tally types and estimators (maybe others too?) no longer agree with the values in the Python interface and it's causing CI to fail.

**Comment by gridley** [comment]
Thanks for ID'ing that issue @pshriwise. Should be an easy fix...

**Comment by gridley** [comment]
OK, so, if you look at the failed test output, it seems that the last displayed digit of some numbers differ but aside from that the difference is like 1e-7, consistently. I have absolutely no clue why this would happen.

**Comment by gridley** [comment]
Oh, OK. It seems that a disadvantage of using enum classes is that you can't change their values through the python ctypes module, which causes the test_settings.py unit test to fail. In particular, changing run_mode of settings.cpp is the only thing which fails. Is this functionality even used? I can only imagine that changing stuff in settings.cpp would have changes made by python immediately over-written once openmc_init gets called and xml-read values overwrite everything.

**Comment by gridley** [comment]
@paulromano I was just hoping to get your feedback on that above comment once you get the chance. Sorry to not ping you earlier.

**Comment by paulromano** [comment]
@gridley The `openmc.lib` module allows you to change things _after_ `openmc_init` is already called. For example, you can add tallies that are not defined in tallies.xml. However, in this particular case (getting/setting the run mode), it doesn't look like it is actually being used anywhere, so I'm personally ok with just removing that binding.

**Comment by gridley** [comment]
OK, thanks. Or we could just change run_mode back to an int and not have to update the CI. Up to you.

**Comment by paulromano** [comment]
Sure, if you keep it as an int, maybe just add a TODO somewhere saying to look into it further (if there isn't one already)

**Pull request closed without merging**